apps/client/src/pricers/uniswapV3/index.ts

```diff
@@ -0,0 +1,153 @@
+import {
+  type Account,
+  type Address,
+  type Chain,
+  type Client,
+  type Transport,
+  erc20Abi,
+  fromHex,
+  zeroAddress,
+} from "viem";
+import { readContract } from "viem/actions";
+
+import type { Pricer } from "../pricer";
+
+import { uniswapV3FactoryAbi, uniswapV3PoolAbi } from "../../abis/uniswapV3";
+import {
+  FEE_TIERS,
+  DEFAULT_FACTORY_ADDRESS,
+  specificFactoryAddresses,
+  USD_REFERENCE,
+} from "@morpho-blue-liquidation-bot/config";
+
+export class UniswapV3Pricer implements Pricer {
+  private pools: Record<Address, Record<Address, Address[]>> = {};
+  private decimals: Record<Address, number> = {};
+
+  supportsChain(chainId: number) {
+    return USD_REFERENCE[chainId] !== undefined;
+  }
+
+  async supportsAsset(client: Client<Transport, Chain, Account>, asset: Address) {
+    // biome-ignore lint/style/noNonNullAssertion: checked before
+    const usdReference = USD_REFERENCE[client.chain.id]!;
+    if (asset === usdReference) return true;
+
+    const pools =
+      this.getCachedPools(asset, usdReference) ??
+      (await this.fetchPools(client, usdReference, asset));
+
+    return pools.length > 0;
+  }
+
+  async price(client: Client<Transport, Chain, Account>, asset: Address) {
+    // biome-ignore lint/style/noNonNullAssertion: checked before
+    const usdReference = USD_REFERENCE[client.chain.id]!;
+
+    /// TODO: allow multiple USD references
+
+    if (asset === usdReference) return 1;
+
+    const pools = this.getCachedPools(asset, usdReference);
+
+    if (pools === undefined) {
+      return;
+    }
+
+    try {
+      const liquidities = await Promise.all(
+        pools.map(async (pool) => {
+          return {
+            pool,
+            amount: await readContract(client, {
+              address: pool,
+              abi: uniswapV3PoolAbi,
+              functionName: "liquidity",
+            }),
+          };
+        }),
+      );
+
+      const biggestPool = liquidities.reduce(
+        (max, liquidity) => (max !== null && liquidity.amount > max.amount ? liquidity : max),
+        liquidities[0] ?? null,
+      )?.pool;
+
+      if (!biggestPool) {
+        throw new Error("No Uniswap pool found");
+      }
+
+      const token0 =
+        fromHex(asset, "bigint") < fromHex(usdReference, "bigint") ? asset : usdReference;
+      const token1 = token0 === asset ? usdReference : asset;
+
+      const [slot0, token0Decimals, token1Decimals] = await Promise.all([
+        readContract(client, {
+          address: biggestPool,
+          abi: uniswapV3PoolAbi,
+          functionName: "slot0",
+        }),
+        this.getDecimals(client, token0),
+        this.getDecimals(client, token1),
+      ]);
+
+      const sqrtPriceX96 = slot0[0];
+      const price =
+        Number((sqrtPriceX96 / 2n ** 96n) ** 2n) / 10 ** (token1Decimals - token0Decimals);
```

Comment:
`(sqrtPriceX96 / 2n ** 96n) ** 2n` won't always fit in a Javascript number. Maybe something like:

```suggestion
        Number(formatUnits(((sqrtPriceX96 / 2n ** 96n) ** 2n) * (10n ** token0Decimals), token1Decimals));
```